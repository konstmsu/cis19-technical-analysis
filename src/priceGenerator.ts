import { linspace } from "./math";
import { sin } from "mathjs";

export function randomSeq(count: number) {
  const result = [];

  for (let i = 0; i < count; i++) {
    result.push(Math.random() * 2 - 1.0);
  }

  return result;
}

export function linearSeq(start: number, stop: number) {
  return (count: number) => linspace(start, stop, count);
}

export function sinSeq(start: number, stop: number) {
  return (count: number): number[] => {
    return linspace(start, stop, count).map(x => sin(x));
  };
}
