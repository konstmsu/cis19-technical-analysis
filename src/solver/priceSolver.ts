export function constantSolver(prices: number[], testSize: number) {
  return [];
}

export function linearSolver(prices: number[], testSize: number) {
  return [0, testSize - 1];
}
