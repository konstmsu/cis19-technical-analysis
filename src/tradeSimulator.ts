export function simulate(prices: number[], trades: number[]) {
  const sortedTrades = [...trades].sort();

  let money = 1.0;
  let equity = 0;
  let holdingEquity = false;

  const buyEquity = (price: number) => {
    equity = money / price;
    money = 0;
    holdingEquity = true;
  };

  function sellEquity(price: number) {
    money = equity * price;
    equity = 0;
    holdingEquity = false;
  }

  for (const t of sortedTrades) {
    const price = prices[t];

    if (holdingEquity) {
      sellEquity(price);
    } else {
      buyEquity(price);
    }
  }

  if (holdingEquity) {
    sellEquity(prices[prices.length - 1]);
  }

  return money;
}
