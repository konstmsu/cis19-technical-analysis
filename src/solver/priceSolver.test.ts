import { linearSeq, randomSeq, sinSeq } from "../priceGenerator";
import { simulate } from "../tradeSimulator";
import { constantSolver, linearSolver, sinSolver } from "./priceSolver";
import { pi } from "mathjs";

function runProblem(
  generatePrices: (count: number) => number[],
  trainSize: number,
  testSize: number,
  solve: (prices: number[], testSize: number) => number[]
) {
  const prices = generatePrices(trainSize + testSize);
  const trainingSet = prices.slice(0, trainSize);
  const testSet = prices.slice(trainSize);
  const trades = solve(trainingSet, testSize);
  return { trainingSet, testSet, trades, amount: simulate(testSet, trades) };
}

describe("price solver", () => {
  it("solve constant", () => {
    const result = runProblem(randomSeq, 10, 100, constantSolver);
    expect(result.amount).toEqual(1.0);
  });

  it("solve linear increasing", () => {
    const result = runProblem(linearSeq(9, 12), 2, 5, linearSolver);
    expect(result.amount).toBeCloseTo(1.2, 5);
  });

  it("solve linear decresing", () => {
    const result = runProblem(linearSeq(12, 7), 2, 4, linearSolver);
    expect(result.amount).toBeCloseTo(1, 5);
  });

  it("sin price seq", () => {
    expect(
      sinSeq(0, 2 * pi * 1.5)(19).map(v => v.toFixed(4))
    ).toMatchSnapshot();
  });

  it("solve sin", () => {
    const sinPriceSeq = (count: number) =>
      sinSeq(0, 2 * pi * 3)(count).map(v => v * 0.5 + 1.5);

    const result = runProblem(sinPriceSeq, 12, 25, sinSolver);
    expect(result.trainingSet.map(v => v.toFixed(4))).toMatchSnapshot(
      "trainingSet"
    );
    expect(result.testSet.map(v => v.toFixed(4))).toMatchSnapshot("testSet");
    expect(result.amount).toBeCloseTo(3, 5);
  });
});
