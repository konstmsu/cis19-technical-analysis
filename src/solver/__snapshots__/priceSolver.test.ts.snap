// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`price solver sin price seq 1`] = `
Array [
  "0.0000",
  "0.5000",
  "0.8660",
  "1.0000",
  "0.8660",
  "0.5000",
  "0.0000",
  "-0.5000",
  "-0.8660",
  "-1.0000",
  "-0.8660",
  "-0.5000",
  "-0.0000",
  "0.5000",
  "0.8660",
  "1.0000",
  "0.8660",
  "0.5000",
  "0.0000",
]
`;
